### PR TITLE
[DomCrawler] Added Crawler::matches(), ::closest(), ::outerHtml()

### LINKED ISSUE #12223: [DomCrawler] Added Crawler::matches(), ::closest(), ::outerHtml()
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#33144
| PR author(s) | @lyrixx
| Merged in    | 4.4

### PR DESCRIPTION
**Pull request by lyrixx**
fixes #12223

### CODE REVIEW

File: components/dom_crawler.rst [hunk 1]
```diff
@@ -165,6 +165,15 @@ Namespaces can be explicitly registered with the
     $crawler->registerNamespace('m', 'http://search.yahoo.com/mrss/');
     $crawler = $crawler->filterXPath('//m:group//yt:aspectRatio');
 
+.. versionadded:: 4.4
+
+    The method ``matches()`` was introduced in Symfony 4.4.
```
Comment (OskarStark) [inline components/dom_crawler.rst]:
```suggestion
    The ``matches()`` method was introduced in Symfony 4.4.
```

File: components/dom_crawler.rst [hunk 2]
```diff
@@ -195,6 +204,14 @@ Get all the direct child nodes matching a CSS selector::
 
     $crawler->filter('body')->children('p.lorem');
 
+.. versionadded:: 4.4
+
+    The method ``closest()`` was introduced in Symfony 4.4.
```
Comment (OskarStark) [inline components/dom_crawler.rst]:
```suggestion
    The ``closest()`` method was introduced in Symfony 4.4.
```

File: components/dom_crawler.rst [hunk 3]
```diff
@@ -337,6 +354,13 @@ and :phpclass:`DOMNode` objects::
 
         The default argument of ``html()`` was introduced in Symfony 4.3.
 
+    .. versionadded:: 4.4
+
+        The method ``outerHtml()`` was introduced in Symfony 4.4.
```
Comment (OskarStark) [inline components/dom_crawler.rst]:
```suggestion
        The ``outerHtml()`` method was introduced in Symfony 4.4.
```

File: components/dom_crawler.rst [hunk 4]
```diff
@@ -195,6 +204,14 @@ Get all the direct child nodes matching a CSS selector::
 
     $crawler->filter('body')->children('p.lorem');
 
+.. versionadded:: 4.4
+
+    The method ``closest()`` was introduced in Symfony 4.4.
+
+Get the first parents (heading toward the document root) of the Element that matches the provided selector.
```
Comment (OskarStark) [inline components/dom_crawler.rst]:
```suggestion
Get the first parents (heading toward the document root) of the Element that matches the provided selector::
```

File: components/dom_crawler.rst [hunk 5]
```diff
@@ -195,6 +204,14 @@ Get all the direct child nodes matching a CSS selector::
 
     $crawler->filter('body')->children('p.lorem');
 
+.. versionadded:: 4.4
```
Comment (OskarStark) [inline components/dom_crawler.rst]: Lets move it below the code example please

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: components/dom_crawler.rst (hunk 5 above)
Comment (lyrixx) [reply components/dom_crawler.rst]: I'm not sure to understand. Is dc1e6b550 ok ?

File: components/dom_crawler.rst [hunk 6]
```diff
@@ -195,6 +203,14 @@ Get all the direct child nodes matching a CSS selector::
 
     $crawler->filter('body')->children('p.lorem');
 
+Get the first parents (heading toward the document root) of the Element that matches the provided selector::
```
Comment (OskarStark) [inline components/dom_crawler.rst]:
```suggestion
Get the first parents (heading toward the document root) of the element that matches the provided selector::
```

File: components/dom_crawler.rst [hunk 7]
```diff
@@ -333,10 +349,18 @@ and :phpclass:`DOMNode` objects::
         // avoid the exception passing an argument that html() returns when node does not exist
         $html = $crawler->html('Default <strong>HTML</strong> content');
 
+        $html = $crawler->outerHtml();
```
Comment (OskarStark) [inline components/dom_crawler.rst]: lets move this below the versionadded of `html()` method, moving it away from the context makes not so much sense and Introduce the outherHtml()` method with a short sentence after the `html()` versionadded.

To be clear regarding the order:

* `html()` code example
* `html()` versionadded directive
* `outherHtml()` introduction/description
* `outherHtml()` code example
* `outherHtml()` versionadded directive

Thanks.

Comment (lyrixx) [reply components/dom_crawler.rst]: Ok got it :+1: thanks.

Is it better now ?

**Review by OskarStark — APPROVED** [review]
Awesome, love it ❤️ Thanks! 👍

**Comment by OskarStark** [comment]
Thanks for your work on this new feature!

**Pull request merged**